### PR TITLE
Implemented all AddNewTransaction validations

### PR DESCRIPTION
**Pull request by electricmonk**

### CODE REVIEW

File: test/crypto/keys/ed25519.go [hunk 1]
```diff
@@ -23,12 +23,12 @@ var ed25519KeyPairs = []ed25519KeyPairHex{
 	{"70d92324eb8d24b7c7ed646e1996f94dcd52934a031935b9ac2d0e5bbcfa357c", "f1c41ba8a1d78f7cdc4f4ff23f3b736e30c630085697d6503e16ac899646f5ab70d92324eb8d24b7c7ed646e1996f94dcd52934a031935b9ac2d0e5bbcfa357c"},
 }
 
-type ed25519KeyPair struct {
+type Ed25519KeyPair struct {
```
Comment (talkol) [inline test/crypto/keys/ed25519.go]: We need to make sure we only use this struct in tests so let's maybe name it explicitly with `Test`?

File: services/transactionpool/add_new_transaction.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		expiryWindow:                30 * time.Minute,
+		lastCommittedBlockTimestamp: primitives.TimestampNano(time.Now().UnixNano()),
+		futureTimestampGrace:        3 * time.Minute,
+		virtualChainId:              primitives.VirtualChainId(42),
+		transactionInPendingPool:    s.isTransactionInPendingPool,
+	}
+	err := validateTransaction(input.SignedTransaction, vctx)
+	if err != nil {
+		s.reporting.Info("transaction is invalid", instrumentation.Error(err), instrumentation.Stringable("transaction", input.SignedTransaction))
+		return nil, err
```
[the remark this answers is not in the document]
Comment (electricmonk) [reply services/transactionpool/add_new_transaction.go]: I think this is dangerous and might lead to confusion (as no one expects both `err` and `out` to be populated as a general principle), but I'll change it for now

File: test/builders/transactions.go [hunk 3]
```diff
@@ -52,55 +52,77 @@ func (t *transaction) Build() *protocol.SignedTransaction {
...
-	t.builder.Transaction.Timestamp = timestamp
+func (t *TransactionBuilder) WithTimestamp(timestamp time.Time) *TransactionBuilder {
+	t.builder.Transaction.Timestamp = primitives.TimestampNano(timestamp.UnixNano())
 	return t
 }
 
-func (t *transaction) WithInvalidSigner(publicKey primitives.Ed25519PublicKey, privateKey primitives.Ed25519PrivateKey) *transaction {
-	corruptPrivateKey := make([]byte, len(privateKey))
-	return t.WithSigner(publicKey, corruptPrivateKey)
+func (t *TransactionBuilder) WithInvalidSigner() *TransactionBuilder {
```
[the remark this answers is not in the document]
Comment (talkol) [reply test/builders/transactions.go]: I'm going to add it back :)

File: test/builders/transactions.go [hunk 4]
```diff
@@ -52,55 +52,77 @@ func (t *transaction) Build() *protocol.SignedTransaction {
...
+func (t *TransactionBuilder) Builder() *protocol.SignedTransactionBuilder {
 	signedTransaction := t.Build()
 	t.builder.Signature = signedTransaction.Signature()
 	return t.builder
 }
 
-func (t *transaction) WithSigner(publicKey primitives.Ed25519PublicKey, privateKey primitives.Ed25519PrivateKey) *transaction {
-	t.builder.Transaction.Signer.Eddsa.SignerPublicKey = publicKey
-	t.signer = privateKey
+func (t *TransactionBuilder) WithSigner(keyPair keys.Ed25519KeyPair) *TransactionBuilder {
```
[the remark this answers is not in the document]
Comment (talkol) [reply test/builders/transactions.go]: Just the other builders (specifically the only one that signs is `benchmarkconsensus`)

**Pull request merged**